lib/dependencies.js

```diff
@@ -0,0 +1,380 @@
+     * @type {Map<string, Array<string>>}
+     **/
+    this.dependencyMap = new Map();
+    /**
+     * Map of filesystem roots to unique dependencies on that filesystem. This
+     * property is only populated once `seal()` is called. This allows us to
+     * build an FSTree (which requires relative paths) per filesystem root.
+     * @type {Map<string, Set<string>>}
+     */
+    this.allDependencies = new Map();
+    /**
+     * Map of filesystem roots to FSTrees, capturing the state of all
+     * dependencies.
+     * @type {Map<string, FSTree<Entry> | FSHashTree>}
+     */
+    this.fsTrees = new Map();
+    /**
+     * Maps dependencies to the files that depend on them.
+     * Keys are absolute paths, values are paths relative to the `rootDir`.
+     * @type Map<string, Array<string>>;
+     */
+    this.dependentsMap = new Map();
+  }
+
+  /**
+   * Seals the dependencies. No more dependencies can be added once this is
+   * called.
+   * @return {this}
+   */
+  seal() {
+    this.sealed = true;
+    this.dependencyMap.forEach((deps, referer) => {
+      for (let i = 0; i < deps.length; i++) {
+        // Build a unified set of dependencies for the entire tree
+        /** @type {string} */
+        let depRoot;
+        if (deps[i].startsWith(this.rootDir + path.sep)) {
+          depRoot = this.rootDir;
+        } else {
+          depRoot = path.parse(deps[i]).root;
+        }
+        let depsForRoot = this._getDepsForRoot(depRoot);
+        depsForRoot.add(path.relative(depRoot, deps[i]));
+
+        // Create an inverse map so that when a dependency is invalidated
+        // we can track it back to the file that should be processed again.
+        let dependents = this.dependentsMap.get(deps[i]);
+        if (!dependents) {
+          dependents = [];
+          this.dependentsMap.set(deps[i], dependents);
+        }
+        dependents.push(referer);
+      }
+    });
+    return this;
+  }
+
+  _getDepsForRoot(dir) {
+    let depsForRoot = this.allDependencies.get(dir);
+    if (!depsForRoot) {
+      depsForRoot = new Set();
+      this.allDependencies.set(dir, depsForRoot);
+    }
+    return depsForRoot;
+  }
+
+  unseal() {
+    this.sealed = false;
+    this.allDependencies.clear();
+    this.fsTrees.clear();
+  }
+
+  countAll() {
+    let num = 0;
+    this.dependencyMap.forEach((deps) => {
+      num += deps.length;
+    });
+    return num;
+  }
+
+  /**
+   * Counts the number of unique dependencies.
+   *
+   * @returns {number}
+   */
+  countUnique() {
+    if (!this.sealed) {
+      throw new Error("Cannot count dependencies until after sealing them.");
+    } else {
+      return this.dependentsMap.size;
+    }
+  }
+
+  /**
+   * Set the dependencies for the file specified by `filePath`.
+   *
+   * @param filePath {string} relative path of the file that has dependencies.
+   * @param dependencies {Array<string>} absolute or relative paths the file
```

Comment:
> At worst, it seems like this code wouldn't have a chance to invalidate the cache because something in the broccoli internals has decided that no invalidation can possibly occur, 

Yes, this is my worry.  As implemented, it is possible for the side-affects of a "out of tree dependency" becoming dirty to not be reflected in the final output.

Reply:
> As implemented

I have tests that show it invaliding cached output in useful situations. What is the condition where this occurs? Is it when the incoming tree has no changes in it's output and so the entire tree is skipped?

If so, we shouldn't let the perfect be the enemy of the good here. This invalidates output that wouldn't be invalidated but should be. It moves us closer to the correct behavior.

Reply:
> If so, we shouldn't let perfect be the enemy of the good here. This invalidates output that wouldn't be invalidated but should be. It moves us closer to the correct behavior.

Ya, after chatting this introducing nothing new in-terms of allowing plugins to escape the default watching system. We can move forward here, and in parallel coordinate with @thoov on next steps at a broccoli level.